File: frontend/src/lib/reportGrouping.ts
/**
 * Report grouping/categorization utilities for the daily report page.
 *
 * Pure functions only — no React, no API calls — so the clubbing and
 * categorization logic can be reasoned about (and unit tested later)
 * independently of rendering.
 */
import type { ReportChangeDetail } from "../types";

// ── Category keywords ────────────────────────────────────────────────────
// Matched against the lowercased file path (substring match, first match
// wins, checked in order). Edit this list as new patch/change types come
// up — nothing else needs to change to add a category.
export const DEFAULT_CATEGORY_KEYWORDS: Array<[label: string, keywords: string[]]> = [
  ["kernel", ["kernel", "/modules/", ".build-id", "vmlinuz", "initramfs"]],
  ["grub", ["grub", "/boot/"]],
  ["httpd", ["httpd"]],
  ["apache", ["apachectl", "apache2"]],
  ["conf", [".conf"]],
];

// Extensions whose files always get individual Mtime detail, regardless of
// how large their category is — config file *content* matters more than
// bulk binary/module counts. This is deliberately extension-based, not
// size-based: a 9-file "kernel" category and a 1-file "apache" category
// don't get individual treatment even though they're just as small as the
// 25-file ".conf" category that does, in the real report this was modeled on.
export const DEFAULT_DETAIL_EXTENSIONS = [".conf", ".cfg", ".yaml", ".yml", ".ini", ".json"];

// Similarity floor (0-1) for clubbing two hosts' change sets together.
// Real patch rollouts are rarely byte-identical across hosts (one extra
// local file, timing differences) — this uses similarity rather than
// requiring an exact match.
export const DEFAULT_CLUB_THRESHOLD = 0.9;

const MAX_SAMPLES_PER_BUCKET = 4;
const ROLLUP_MIN_COUNT = 20; // only worth a directory rollup line above this many files

/**
 * Collapse multiple changes for the same (host, file_path) down to just the
 * most recent one, by current_mtime. The same file can legitimately show up
 * more than once in a single day's report — e.g. /etc/shadow changing twice
 * in one day from separate password resets — and every downstream count
 * (bucket totals, "N changes" headers, clubbing similarity) should reflect
 * one logical change per file, not each individual detection.
 */
export function dedupeByLatestMtime(changes: ReportChangeDetail[]): ReportChangeDetail[] {
  const latest = new Map<string, ReportChangeDetail>();
  for (const c of changes) {
    const key = `${c.agent_hostname ?? ""}:${normalizedType(c)}:${c.file_path}`;
    const existing = latest.get(key);
    if (!existing) {
      latest.set(key, c);
      continue;
    }
    const existingTime = existing.current_mtime ? Date.parse(existing.current_mtime) : -Infinity;
    const candidateTime = c.current_mtime ? Date.parse(c.current_mtime) : -Infinity;
    if (candidateTime >= existingTime) latest.set(key, c);
  }
  return Array.from(latest.values());
}

function normalizedType(c: ReportChangeDetail): string {
  return (c.change_type || "").toLowerCase();
}

export function categorizeFile(
  path: string,
  keywords: Array<[string, string[]]> = DEFAULT_CATEGORY_KEYWORDS,
): string {
  const lower = path.toLowerCase();
  for (const [label, kws] of keywords) {
    if (kws.some(kw => lower.includes(kw))) return label;
  }
  return "other";
}

export function isDetailWorthy(
  path: string,
  extensions: string[] = DEFAULT_DETAIL_EXTENSIONS,
): boolean {
  const lower = path.toLowerCase();
  return extensions.some(ext => lower.endsWith(ext));
}

export interface CategoryBucket {
  category: string;
  count: number;
  samples: string[];
  moreCount: number;
}

/** Group changes of one change_type ("added"/"removed"/"changed") into category buckets. */
export function buildBuckets(
  changes: ReportChangeDetail[],
  changeType: string,
  keywords: Array<[string, string[]]> = DEFAULT_CATEGORY_KEYWORDS,
): CategoryBucket[] {
  const relevant = changes.filter(c => normalizedType(c) === changeType);
  const byCategory = new Map<string, ReportChangeDetail[]>();
  for (const c of relevant) {
    const cat = categorizeFile(c.file_path, keywords);
    if (!byCategory.has(cat)) byCategory.set(cat, []);
    byCategory.get(cat)!.push(c);
  }

  const buckets: CategoryBucket[] = [];
  for (const [category, items] of byCategory.entries()) {
    const samples = items.slice(0, MAX_SAMPLES_PER_BUCKET).map(c => c.file_path);
    buckets.push({
      category,
      count: items.length,
      samples,
      moreCount: Math.max(0, items.length - samples.length),
    });
  }

  // Largest first; "other" always sorts last regardless of size, since
  // it's a catch-all rather than a meaningful pattern.
  return buckets.sort((a, b) => {
    if (a.category === "other") return 1;
    if (b.category === "other") return -1;
    return b.count - a.count;
  });
}

export interface DirectoryRollup {
  directory: string;
  count: number;
}

/** Roll up large added/removed subtrees into one line instead of listing every file. */
export function buildDirectoryRollups(
  changes: ReportChangeDetail[],
  changeType: string,
): DirectoryRollup[] {
  const relevant = changes.filter(c => normalizedType(c) === changeType);
  const byDir = new Map<string, number>();
  for (const c of relevant) {
    const idx = c.file_path.lastIndexOf("/");
    const dir = idx > 0 ? c.file_path.slice(0, idx + 1) : "/";
    byDir.set(dir, (byDir.get(dir) || 0) + 1);
  }
  return Array.from(byDir.entries())
    .filter(([, count]) => count >= ROLLUP_MIN_COUNT)
    .map(([directory, count]) => ({ directory, count }))
    .sort((a, b) => b.count - a.count);
}

/** Changed files worth showing individually (see DEFAULT_DETAIL_EXTENSIONS doc above). */
export function buildDetailEntries(
  changes: ReportChangeDetail[],
  extensions: string[] = DEFAULT_DETAIL_EXTENSIONS,
): ReportChangeDetail[] {
  return changes.filter(
    c => normalizedType(c) === "changed"
      && isDetailWorthy(c.file_path, extensions)
      && (c.baseline_mtime || c.current_mtime),
  );
}

// ── Host clubbing ─────────────────────────────────────────────────────────

function changeFingerprint(c: ReportChangeDetail): string {
  // Same file + same change type + same resulting hash = the same actual
  // change, not just coincidentally the same path (e.g. /etc/mtab changes
  // on every host with a different hash each time — that must NOT count
  // as a shared pattern).
  return `${normalizedType(c)}:${c.file_path}:${c.current_hash ?? ""}`;
}

/** Jaccard similarity between two hosts' change sets, by fingerprint. */
export function computeSimilarity(a: ReportChangeDetail[], b: ReportChangeDetail[]): number {
  if (a.length === 0 && b.length === 0) return 1;
  const setA = new Set(a.map(changeFingerprint));
  const setB = new Set(b.map(changeFingerprint));
  let intersection = 0;
  for (const key of setA) if (setB.has(key)) intersection++;
  const union = setA.size + setB.size - intersection;
  return union === 0 ? 1 : intersection / union;
}

export interface HostChanges {
  hostname: string;
  changes: ReportChangeDetail[];
}

export interface HostGroup {
  hostnames: string[];
  changes: ReportChangeDetail[]; // union of every member host's changes, deduped by fingerprint
}

/**
 * Greedy single-pass clustering: each unclubbed host becomes a seed, and
 * any remaining host similar enough to *that seed* joins its group. This
 * compares against the seed only (not the growing merged set) — a
 * deliberate simplification that's more than adequate at the scale of a
 * handful to a few dozen hosts per report, and avoids O(n^2) recompute on
 * a growing set.
 */
export function clubHosts(
  hosts: HostChanges[],
  threshold: number = DEFAULT_CLUB_THRESHOLD,
): { groups: HostGroup[]; solos: HostChanges[] } {
  // Dedupe first — a file detected twice in one day for one host must not
  // inflate that host's change-set size or skew its similarity to others.
  const remaining = hosts.map(h => ({ ...h, changes: dedupeByLatestMtime(h.changes) }));
  const groups: HostGroup[] = [];
  const solos: HostChanges[] = [];

  while (remaining.length > 0) {
    const seed = remaining.shift()!;
    const members = [seed];

    for (let i = remaining.length - 1; i >= 0; i--) {
      const candidate = remaining[i];
      if (computeSimilarity(seed.changes, candidate.changes) >= threshold) {
        members.push(candidate);
        remaining.splice(i, 1);
      }
    }

    if (members.length >= 2) {
      const seen = new Set<string>();
      const merged: ReportChangeDetail[] = [];
      for (const m of members) {
        for (const c of m.changes) {
          const key = changeFingerprint(c);
          if (!seen.has(key)) {
            seen.add(key);
            merged.push(c);
          }
        }
      }
      groups.push({ hostnames: members.map(m => m.hostname), changes: merged });
    } else {
      solos.push(seed);
    }
  }

  return { groups, solos };
}

File: app/services/report_grouping.py
"""
Report grouping/categorization -- Python port of
frontend/src/lib/reportGrouping.ts, so the RT-published report (built here,
server-side) shows the same category buckets / host-clubbing / directory
rollups the on-screen report UI does, instead of a flat per-file listing.

Keep this in sync with reportGrouping.ts by hand -- there's no shared
source of truth between the two runtimes. Constants and algorithms below
are intentionally identical to that file; see its own comments for the
full design rationale (project memory "report-grouping-design-pending").

Each `changes` list here is a list of dicts shaped like the per-change
dicts app/api/reports.py's _build_publish_agents_data builds: file_path,
change_type, severity, baseline_hash, current_hash, baseline_size,
current_size, analyst_notes, is_known_change, requires_investigation,
baseline_mtime, current_mtime, audit_uid, audit_process, audit_command
(mtimes as ISO strings or None).
"""
from typing import Dict, List, Optional, Tuple

# ── Category keywords ────────────────────────────────────────────────────
DEFAULT_CATEGORY_KEYWORDS: List[Tuple[str, List[str]]] = [
    ("kernel", ["kernel", "/modules/", ".build-id", "vmlinuz", "initramfs"]),
    ("grub", ["grub", "/boot/"]),
    ("httpd", ["httpd"]),
    ("apache", ["apachectl", "apache2"]),
    ("conf", [".conf"]),
]

DEFAULT_DETAIL_EXTENSIONS = [".conf", ".cfg", ".yaml", ".yml", ".ini", ".json"]

DEFAULT_CLUB_THRESHOLD = 0.9

MAX_SAMPLES_PER_BUCKET = 4
ROLLUP_MIN_COUNT = 20
MAX_GROUP_HOSTS_SHOWN = 50
MIN_ROLLUP_DEPTH = 2


def _normalized_type(c: Dict) -> str:
    return (c.get("change_type") or "").lower()


def dedupe_by_latest_mtime(changes: List[Dict]) -> List[Dict]:
    """Collapse repeated (change_type, file_path) entries down to the most
    recent by current_mtime -- same file can legitimately show up more than
    once in a day (e.g. two separate password resets touching /etc/shadow)."""
    latest: Dict[str, Dict] = {}
    for c in changes:
        key = f"{_normalized_type(c)}:{c.get('file_path')}"
        existing = latest.get(key)
        if existing is None:
            latest[key] = c
            continue
        # ISO 8601 strings sort lexicographically the same as chronologically.
        if (c.get("current_mtime") or "") >= (existing.get("current_mtime") or ""):
            latest[key] = c
    return list(latest.values())


def categorize_file(path: str, keywords: List[Tuple[str, List[str]]] = DEFAULT_CATEGORY_KEYWORDS) -> str:
    lower = path.lower()
    for label, kws in keywords:
        if any(kw in lower for kw in kws):
            return label
    return "other"


def is_detail_worthy(path: str, extensions: List[str] = DEFAULT_DETAIL_EXTENSIONS) -> bool:
    lower = path.lower()
    return any(lower.endswith(ext) for ext in extensions)


def build_buckets(
    changes: List[Dict], change_type: str,
    keywords: List[Tuple[str, List[str]]] = DEFAULT_CATEGORY_KEYWORDS,
) -> List[Dict]:
    """Group one change_type's changes into category buckets."""
    relevant = [c for c in changes if _normalized_type(c) == change_type]
    by_category: Dict[str, List[Dict]] = {}
    for c in relevant:
        cat = categorize_file(c.get("file_path") or "", keywords)
        by_category.setdefault(cat, []).append(c)

    buckets = []
    for category, items in by_category.items():
        samples = [c.get("file_path") for c in items[:MAX_SAMPLES_PER_BUCKET]]
        buckets.append({
            "category": category,
            "count": len(items),
            "samples": samples,
            "more_count": max(0, len(items) - len(samples)),
        })

    # Largest first; "other" always last regardless of size.
    buckets.sort(key=lambda b: (b["category"] == "other", -b["count"]))
    return buckets


def _parent_of(directory: str) -> Optional[str]:
    """One level up from `directory` (which always ends in '/'), or None if
    that would go above MIN_ROLLUP_DEPTH."""
    trimmed = directory[:-1] if directory.endswith("/") else directory
    idx = trimmed.rfind("/")
    if idx <= 0:
        return None
    parent = trimmed[: idx + 1]
    depth = len([seg for seg in parent.split("/") if seg])
    return parent if depth >= MIN_ROLLUP_DEPTH else None


def _merge_sibling_rollups(entries: Dict[str, int]) -> Dict[str, int]:
    """Merge 2+ sibling directory rollups sharing a parent into one combined
    line at that parent, repeating until a pass produces no further merges.
    Always terminates -- _parent_of is bounded by MIN_ROLLUP_DEPTH."""
    current = dict(entries)
    while True:
        by_parent: Dict[str, List[str]] = {}
        for directory in current:
            parent = _parent_of(directory)
            if parent is None:
                continue
            by_parent.setdefault(parent, []).append(directory)

        next_entries = dict(current)
        merged_any = False
        for parent, dirs in by_parent.items():
            if len(dirs) < 2:
                continue
            total = sum(current.get(d, 0) for d in dirs)
            for d in dirs:
                next_entries.pop(d, None)
            next_entries[parent] = next_entries.get(parent, 0) + total
            merged_any = True

        current = next_entries
        if not merged_any:
            return current


def build_directory_rollups(changes: List[Dict], change_type: str) -> List[Dict]:
    """Roll up large added/removed subtrees into one line instead of listing
    every file, merging sibling subdirectories under a shared parent."""
    relevant = [c for c in changes if _normalized_type(c) == change_type]
    by_dir: Dict[str, int] = {}
    for c in relevant:
        path = c.get("file_path") or ""
        idx = path.rfind("/")
        directory = path[: idx + 1] if idx > 0 else "/"
        by_dir[directory] = by_dir.get(directory, 0) + 1

    # Only already-qualifying directories are merge candidates -- don't sum
    # small unrelated subdirectories together just because their total
    # happens to cross the threshold.
    qualifying = {d: n for d, n in by_dir.items() if n >= ROLLUP_MIN_COUNT}
    merged = _merge_sibling_rollups(qualifying)

    return sorted(
        ({"directory": d, "count": n} for d, n in merged.items()),
        key=lambda r: -r["count"],
    )


def build_detail_entries(
    changes: List[Dict], extensions: List[str] = DEFAULT_DETAIL_EXTENSIONS,
) -> List[Dict]:
    """Changed files worth showing individually: matches a detail-worthy
    extension, or carries auditd attribution regardless of extension
    (critical-path files like /etc/passwd have no matching extension but
    attribution is exactly the kind of detail this section exists for)."""
    return [
        c for c in changes
        if _normalized_type(c) == "changed"
        and (
            is_detail_worthy(c.get("file_path") or "", extensions)
            or bool(c.get("audit_uid") or c.get("audit_process") or c.get("audit_command"))
        )
        and (c.get("baseline_mtime") or c.get("current_mtime"))
    ]


# ── Host clubbing ─────────────────────────────────────────────────────────

def _change_fingerprint(c: Dict) -> str:
    """Path + change type only -- hash deliberately excluded (see
    reportGrouping.ts's changeFingerprint for the full reasoning: analysts
    club by "did the same file change," not byte-identical content)."""
    return f"{_normalized_type(c)}:{c.get('file_path')}"


def compute_similarity(a: List[Dict], b: List[Dict]) -> float:
    """Jaccard similarity between two hosts' change sets, by fingerprint."""
    if not a and not b:
        return 1.0
    set_a = {_change_fingerprint(c) for c in a}
    set_b = {_change_fingerprint(c) for c in b}
    intersection = len(set_a & set_b)
    union = len(set_a) + len(set_b) - intersection
    return 1.0 if union == 0 else intersection / union


def club_hosts(hosts: List[Dict], threshold: float = DEFAULT_CLUB_THRESHOLD) -> Dict[str, List[Dict]]:
    """
    Greedy single-pass clustering, matching reportGrouping.ts's clubHosts:
    each unclubbed host becomes a seed, and any remaining host similar
    enough to *that seed* (not the growing merged set) joins its group.
    `hosts`: [{"hostname": str, "changes": [...]}, ...].
    Returns {"groups": [{"hostnames": [...], "changes": [...]}], "solos": [{"hostname":..., "changes":...}]}.
    """
    remaining = [
        {"hostname": h["hostname"], "changes": dedupe_by_latest_mtime(h["changes"])}
        for h in hosts
    ]
    groups: List[Dict] = []
    solos: List[Dict] = []

    while remaining:
        seed = remaining.pop(0)
        members = [seed]

        i = len(remaining) - 1
        while i >= 0:
            candidate = remaining[i]
            if compute_similarity(seed["changes"], candidate["changes"]) >= threshold:
                members.append(candidate)
                remaining.pop(i)
            i -= 1

        if len(members) >= 2:
            seen = set()
            merged: List[Dict] = []
            for m in members:
                for c in m["changes"]:
                    key = _change_fingerprint(c)
                    if key not in seen:
                        seen.add(key)
                        merged.append(c)
            groups.append({"hostnames": [m["hostname"] for m in members], "changes": merged})
        else:
            solos.append(seed)

    return {"groups": groups, "solos": solos}
